src/NuGet.Services.Metadata.Catalog.Monitoring/Status/PackageMonitoringStatusService.cs

```diff
@@ -70,60 +97,68 @@ public PackageMonitoringStatusService(IStorageFactory storageFactory, ILogger<Pa
 
             var storage = GetStorage(state);
 
-            var statusTasks = 
+            var statusTasks =
                 (await storage.List(token))
-                .Select(listItem =>
-                    Task.Run(async () =>
-                    {
-                        return await GetPackageAsync(storage, listItem.Uri, token);
-                    })
-                );
+                .Select(listItem => GetPackageAsync(storage, listItem.Uri, token))
+                .ToList();
 
-            return 
+            return
                 (await Task.WhenAll(statusTasks))
                 .Where(s => s != null);
         }
-        
+
         public async Task UpdateAsync(PackageMonitoringStatus status, CancellationToken token)
         {
-            // Guarantee that we never have the same package in multiple states by deleting it first.
-            await DeleteAsync(status.Package, token);
+            foreach (int stateInt in Enum.GetValues(typeof(PackageState)))
+            {
+                var state = (PackageState)stateInt;
+                if (state == status.State)
+                {
+                    // Save the new status.
+                    await SaveAsync(status, token);
+                }
+                else
+                {
+                    // Delete the existing status.
+                    await DeleteAsync(status.Package, state, token);
```

Comment:
What would be the state id the DeleteAsync fails ? Will be future update operation impacted?

Reply:
This method would throw, the queue message for this package to be validated would not be removed from the queue, and the validation would run again in approximately 5 minutes.